From 5b056349dc90ae2f974ab04a2b3ff7c3e1205775 Mon Sep 17 00:00:00 2001
Message-ID: <5b056349dc90ae2f974ab04a2b3ff7c3e1205775.1773161960.git.teddy.astie@vates.tech>
In-Reply-To: <b8078ed45d66b94467b594ac7e8d849af32a0984.1773161960.git.teddy.astie@vates.tech>
References: <b8078ed45d66b94467b594ac7e8d849af32a0984.1773161960.git.teddy.astie@vates.tech>
From: Teddy Astie <teddy.astie@vates.tech>
Date: Mon, 9 Mar 2026 10:57:06 +0100
Subject: [PATCH 10/15] x86/vpmu: Expose up to 8 Intel event selectors in PV
 Dom0

Most (if not all) Intel CPUs starting from Sandy Bridge have
up to 8 event selectors per core, which could be halved per
hyperthread.

However, current PV emulation logic doesn't support up to 8
event selector, leading to errors when trying to access them,
hence, preventing Linux from driving the vPMU correctly.

Make sure up to MSR_P6_EVNTSEL(7) is usable, which is the same
upper bound as used in VMX code.

The check if the event selector actually exist for the hardware
is done in core2_vpmu_do_{rdmsr,wrmsr}, hence we're not allowing
to access non-existent MSRs.

Fixes: 27c554198666 ("x86/VPMU: add support for PMU register handling on PV guests")
Signed-off-by: Teddy Astie <teddy.astie@vates.tech>
---
 xen/arch/x86/pv/emul-priv-op.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/xen/arch/x86/pv/emul-priv-op.c b/xen/arch/x86/pv/emul-priv-op.c
index c9f866781c..c53552203b 100644
--- a/xen/arch/x86/pv/emul-priv-op.c
+++ b/xen/arch/x86/pv/emul-priv-op.c
@@ -983,7 +983,7 @@ static int cf_check read_msr(
         return X86EMUL_OKAY;
 
     case MSR_P6_PERFCTR(0) ... MSR_P6_PERFCTR(7):
-    case MSR_P6_EVNTSEL(0) ... MSR_P6_EVNTSEL(3):
+    case MSR_P6_EVNTSEL(0) ... MSR_P6_EVNTSEL(7):
     case MSR_CORE_PERF_FIXED_CTR0 ... MSR_CORE_PERF_FIXED_CTR2:
     case MSR_CORE_PERF_FIXED_CTR_CTRL ... MSR_CORE_PERF_GLOBAL_OVF_CTRL:
         if ( boot_cpu_data.x86_vendor == X86_VENDOR_INTEL )
@@ -1160,7 +1160,7 @@ static int cf_check write_msr(
         break;
 
     case MSR_P6_PERFCTR(0) ... MSR_P6_PERFCTR(7):
-    case MSR_P6_EVNTSEL(0) ... MSR_P6_EVNTSEL(3):
+    case MSR_P6_EVNTSEL(0) ... MSR_P6_EVNTSEL(7):
     case MSR_CORE_PERF_FIXED_CTR0 ... MSR_CORE_PERF_FIXED_CTR2:
     case MSR_CORE_PERF_FIXED_CTR_CTRL ... MSR_CORE_PERF_GLOBAL_OVF_CTRL:
         if ( boot_cpu_data.x86_vendor == X86_VENDOR_INTEL )
-- 
2.53.0

